### PR TITLE
Update code to comply with WP Coding Standards (rebased)

### LINKED ISSUE #8: Set up project in Composer

### LINKED ISSUE #10: Ensure that code is compliant with WP Coding Standards

### LINKED ISSUE #11: Add PHP_CodeSniffer pre-commit hook

### PR DESCRIPTION
**Pull request by Rumperuu**
Updates the codebase to ensure compliance with the WP Coding Standards.

Also Composerises the project to allow for the installation of a pre-commit PHP_CodeSniffer hook (and other hooks later on).

Note 1: because pretty much every line of every file is being changed (even if just to change indentation format), the diff tool is not particularly useful on this PR.

Note 2: This branch's commit history will be squashed on merge, so do not make any commits to this branch that you're going to want to remain in the project history. Keep it scoped to WP Coding Standards compliance only.

**This is a duplicate of #23, except rebased to separate commit https://github.com/markcheret/footnotes/pull/23/commits/3a44a194994763ef2f6370f1cd3aa0d7d22f8a7e into two different commits (https://github.com/markcheret/footnotes/commit/27abf833941a2a5921f71e79f7248a57876210b5 and https://github.com/markcheret/footnotes/commit/b349544fc26b2f2608a6a03d782f0ebf9f26fb3a), ensuring the files in `class/dashboard/*` are properly tracked as having been renamed to `class/layout/*`, rather than just deleted.**

Replace #23 

Close #8
Close #10
Close #11

### CODE REVIEW

**Comment by Rumperuu** [comment]
Reposting @pewgeuges's comment from #23:

**Comment by pewgeuges** [comment]
Thanks Rumperuu for re-posting.
The deleted comment from Rumperuu was only my feedback by email sent to the previous branch #23 and that I deleted there after changing my mind. Meanwhile mirrored here, hence deleted again as out of sync with the ongoing review.

**Comment by Rumperuu** [comment]
Tried rebasing to the new `main` but it seems to have gone a bit squiffy. No damage done, just an ugly patch of commit history.

**Comment by Rumperuu** [comment]
Ah! I was trying to be too clever with all of my double-wildcarding and accidentally ignored the `widget/` and `layout/` directories :man_facepalming: 

Command updated in README and those files are now ALMOST fixed - just need to add nonce verification now

File: class/config.php [hunk 1]
```diff
@@ -1,83 +1,76 @@
...
+ *
+ * Last modified: 2021-02-18T2027+0100
+ */
+
+/**
+ * Contains all Plugin Constants. Contains no Method or Property.
+ *
+ * @since 1.5.0
+ */
+class MCI_Footnotes_Config {
```
Comment (Rumperuu) [inline class/config.php:20]: Hey @pewgeuges, line 20 of `class/config.php` is where the party is!

Comment (Rumperuu) [reply class/config.php:20]: :partying_face:

**Comment by Rumperuu** [comment]
> Ah! I was trying to be too clever with all of my double-wildcarding and accidentally ignored the `widget/` and `layout/` directories man_facepalming
> 
> Command updated in README and those files are now ALMOST standards-compliance now - just need to add nonce verification.
> 
> Still getting this error when activating the Plugin though, so fixing that first:
> 
> > Fatal error: Uncaught Error: Class 'MCI_Footnotes_Layout_Engine' not found in /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/class/layout/diagnostics.php:15 Stack trace: #0 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/includes.php(32): require_once() #1 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/includes.php(38): mci_footnotes_require_php_files('/srv/www/wordpr...') #2 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/footnotes.php(55): require_once('/srv/www/wordpr...') #3 /srv/www/wordpress-one/public_html/wp-admin/includes/plugin.php(2299): include('/srv/www/wordpr...') #4 /srv/www/wordpress-one/public_html/wp-admin/plugins.php(191): plugin_sandbox_scrape('footnotes/footn...') #5 {main} thrown in /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/class/layout/diagnostics.php on line 15

Broken as of bb3b555

**Comment by Rumperuu** [comment]
> > Ah! I was trying to be too clever with all of my double-wildcarding and accidentally ignored the `widget/` and `layout/` directories man_facepalming
> > Command updated in README and those files are now ALMOST standards-compliance now - just need to add nonce verification.
> > Still getting this error when activating the Plugin though, so fixing that first:
> > > Fatal error: Uncaught Error: Class 'MCI_Footnotes_Layout_Engine' not found in /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/class/layout/diagnostics.php:15 Stack trace: #0 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/includes.php(32): require_once() #1 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/includes.php(38): mci_footnotes_require_php_files('/srv/www/wordpr...') #2 /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/footnotes.php(55): require_once('/srv/www/wordpr...') #3 /srv/www/wordpress-one/public_html/wp-admin/includes/plugin.php(2299): include('/srv/www/wordpr...') #4 /srv/www/wordpress-one/public_html/wp-admin/plugins.php(191): plugin_sandbox_scrape('footnotes/footn...') #5 {main} thrown in /srv/www/wordpress-one/public_html/wp-content/plugins/footnotes/class/layout/diagnostics.php on line 15
> 
> Broken as of [bb3b555](https://github.com/markcheret/footnotes/commit/bb3b55514829dda0e61f038e8c8844c3ae656d86)

Fixed in 6eda55d

After painstakingly comparing every diff from bb3b555 line-by-line, I realised that the reason it was failing is because `includes.php` `require_once()`s files alphabetically per-directory, so by renaming `layout.php` to `engine.php` and `subpage-diagnostics.php` to `diagnostics.php` I can accidentally thrown off the order. Temporarily fixed by renaming `engine.php` to `abstract-engine.php`, but really we should refactor `includes.php` to be smarter when we get the chance.

**Comment by pewgeuges** [comment]
Congrats for the fix!

**Review by markcheret — APPROVED** [review]
thanks for adding all contributors back in

File: class/task.php [hunk 2]
```diff
@@ -1066,7 +1066,7 @@ public function the_content( $p_str_content ) {
 	 * @return string Content with replaced footnotes.
 	 */
 	public function the_excerpt( $p_str_content ) {
-		return $this->exec( $p_str_content, false, ! ! MCI_Footnotes_Settings::instance()->get( MCI_Footnotes_Settings::C_BOOL_FOOTNOTES_IN_EXCERPT ) );
+		return $this->exec( $p_str_content, false, ! MCI_Footnotes_Convert::to_bool( MCI_Footnotes_Settings::instance()->get( MCI_Footnotes_Settings::C_BOOL_FOOTNOTES_IN_EXCERPT ) ) );
```
Comment (Rumperuu) [inline class/task.php:1069]: @pewgeuges Will leave this for now (as it's arguably outside the scope of this branch), but I'll explain what I was doing and we can sort it out later:

`MCI_Footnotes_Convert::to_bool` is not necessary, as PHP comes with type conversion (`(bool) <variable>`) built-in. The double-not-operator (`! !`) is a shorthand was of saying that, as the first `!` automatically converts the variable into a `bool` if it was not already, and then the second `!` undoes the negation. So for example:

```
$foo = 'true';
$bar = ! $foo; // $bar === false, as 'true' is a truthy value and automatically cast to `false`
$bar = ! ! $foo; // $bar === true, as 'true' is cast to `true`, negated to `false` and then negated again back to `true`
```
So `$bar = ! ! $foo` is the same as `$bar = (bool) $foo` which is the same as `$bar = MCI_Footnotes_Convert::to_bool($foo)`. More info [here](https://stackoverflow.com/questions/2127260/double-not-operator-in-php).

On reflection, `(bool) $foo` is the most obvious so we should refactor `MCI_Footnotes_Convert::to_bool()` out in future, but no need to now as it functions fine, it's just unnecessary.

**Comment by Rumperuu** [comment]
Added note to PR body about commit squash after I [started going out-of-scope](https://github.com/markcheret/footnotes/pull/32#discussion_r580492899).

File: class/task.php (hunk 2 above)
Comment (Rumperuu) [reply class/task.php:1069]: The advantage of using `(bool)` is doubly obvious considering the original test was `! MCI_Footnotes_Convert::to_bool()` so I should have only had the one `!` !

Comment (pewgeuges) [reply class/task.php:1069]: Thanks @Rumperuu! Sorry for this commit.

File: class/init.php [hunk 3]
```diff
@@ -206,16 +206,22 @@ public function register_public() {
...
+			 *
+			 * Deferring to the footer breaks jQuery tooltip display.
+			 * @date 2021-02-23T1105+0100
 			 */
 			wp_enqueue_script(
 				'mci-footnotes-jquery-tools',
 				plugins_url( 'footnotes/js/jquery.tools.min.js' ),
 				array(),
 				'1.2.7.redacted.2',
-				true
```
Comment (Rumperuu) [inline class/init.php]: Good spot @pewgeuges!

File: class/init.php [hunk 4]
```diff
@@ -200,53 +194,25 @@ public function register_public() {
...
-			 * not use '-js' in the handle, is appended automatically
+			 * not use '-js' in the handle, is appended automatically.
 			 */
 			wp_enqueue_script(
 				'mci-footnotes-jquery-tools',
 				plugins_url( 'footnotes/js/jquery.tools.min.js' ),
 				array(),
-				'1.2.7.redacted.2'
+				'1.2.7.redacted.2',
+				true
```
Comment (Rumperuu) [inline class/init.php]: This should be `false`, fixed by @pewgeuges in 9226a40

**Review by markcheret — APPROVED** [review]
This is only a test comment. Thanks for your work

**Comment by markcheret** [comment]
> This is only a test comment. Thanks for your work

another test

**Review by pewgeuges — CHANGES_REQUESTED** [review]
This GitHub thing is totally unusable on this branch. The plethora of mostly duplicated commits due to bugs on the platform is hard to review in detail, while the changed files view is worthless. It just shows all files with changes in them, and is unable to detect the many identical strings. @Rumperuu did not move blocks of code, so the useless diff view is the sole responsibility of the platform. Note that all space indentation was already converted back to tabs! So the code is fully diffable, but the platform is failing.

As we must track deleted docblocks and disruptive changes to the code. I cannot approve until I’ve diffed all changed files locally in VSCode.

That will take some time, and over this we should add all the missing docblocks as well, since we’re on it and so much is already fixed (or broken).

File: class/init.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rumperuu) [reply class/init.php]: Fixed in 9226a40

File: class/task.php [hunk 5]
```diff
@@ -160,59 +162,68 @@ class MCI_Footnotes_Task {
...
-	 * @reporter @psykonevro
-	 * @link https://wordpress.org/support/topic/footnotes-is-not-amp-compatible/
+	 * @since 2.0.0
+	 * @var str
+	 */
+	private static $a_str_link_ids_separator = '+';
+
+	/**
+	 * Contains the concatenated link.
 	 *
```
[the remark this answers is not in the document]
Comment (Rumperuu) [reply class/task.php:255]: Should have been clearer; they don't need to be `public` class variables as they are never used outside of this class (and `private` class variables don't need docblock comments, so we can return to the previous grouping of variables under a single explanatory comment — adding temporary docblocks to each variable was needed to pass the WP Coding Standard check)

**Comment by Rumperuu** [comment]
> This GitHub thing is totally unusable on this branch. The plethora of mostly duplicated commits due to bugs on the platform is hard to review in detail, while the changed files view is worthless. It just shows all files with changes in them, and is unable to detect the many identical strings. @Rumperuu did not move blocks of code, so the useless diff view is the sole responsibility of the platform. Note that all space indentation was already converted back to tabs! So the code is fully diffable, but the platform is failing.
> 
> As we must track deleted docblocks and disruptive changes to the code, I cannot approve until I’ve diffed all changed files locally in VSCode.
> 
> That will take some time, and over this we should add all the missing docblocks as well, since we’re on it and so much is already fixed (or broken).

You're right - the indentation isn't the issue. It's newlines. Will update in a sec

**Comment by markcheret** [comment]
Do you need to rebase again?

**Comment by Rumperuu** [comment]
> Do you need to rebase again?

It's working locally now:

```
rumperuu@ben-desktop:footnotes$ git diff main includes.php
diff --git a/includes.php b/includes.php
index 732cfb9..9501097 100644
--- a/includes.php
+++ b/includes.php
@@ -3,36 +3,37 @@
  * Includes all common files.
  *
  * @filesource
- * @author Stefan Herndler
+ * @package footnotes
  * @since 1.5.0 14.09.14 13:40
```

I think the GitHub one might have some caching so it might take a while to update.

**Comment by Rumperuu** [comment]
> > Do you need to rebase again?
> 
> It's working locally now:
> 
> ```
> rumperuu@ben-desktop:footnotes$ git diff main includes.php
> diff --git a/includes.php b/includes.php
> index 732cfb9..9501097 100644
> --- a/includes.php
> +++ b/includes.php
> @@ -3,36 +3,37 @@
>   * Includes all common files.
>   *
>   * @filesource
> - * @author Stefan Herndler
> + * @package footnotes
>   * @since 1.5.0 14.09.14 13:40
> ```
> 
> I think the GitHub one might have some caching so it might take a while to update.

Not a caching issue; I think it's still diffing this branch with the version of `main` that it was branched from rather than the current version (with the line endings), so rather than mess around trying to rebase again I've just recreated the branch and created a new PR at #38 

Closing this PR

**Pull request closed without merging**